merkle-tree-incremental/src/Crypto/Hash/MerkleTreeIncremental/Constructor.hs

```diff
@@ -0,0 +1,39 @@
+module Crypto.Hash.MerkleTreeIncremental.Constructor (
+    Constructor,
+    empty,
+    add,
+    finalize,
+)
+where
+
+import Crypto.Hash (HashAlgorithm)
+import Crypto.Hash.MerkleTreeIncremental (MerkleHash, MerkleTree (MerkleTreeEmpty, MerkleTreeRoot), leafHash, nodeHash)
+import Data.ByteArray (ByteArrayAccess)
+
+newtype Constructor a = Constructor [ConstructorNode a]
```

Comment:
Maybe something less generic, for example `MerkleTreeState` like in Blake2 libraries: 

https://hackage.haskell.org/package/blake2-0.3.0.1/docs/Crypto-Hash-BLAKE2-BLAKE2bp.html